Add support for simultaneous runs of Script helper - Part 2

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a follow-on to #31937.

While working to use those changes in the script integration a few bugs were discovered. Also it became clear that a slight change of direction was needed regarding how to configure/handle long running scripts. Lastly a new `queue` mode was added which can be very helpful in certain scenarios (especially for automation action sequences.)

The changes were broken up into several, smaller commits to help make the review process easier.

### New Script helper option

> Replaces `if_running`.

#### script_mode
- Controls what happens if Script run (i.e., `async_run` called) while previous run has not yet completed.
- Values:
  - `error` Raise an exception
  - `ignore` Return without doing anything (previous run continues as-is)
  - `legacy` Implements previous behavior, which is to return when done, or when suspended by `delay` or `wait_template`, whichever comes first. This is the default
  - `parallel` Start run in new task
  - `queue` Start new run when previous runs have completed
  - `restart` Stop previous run before starting new run

### New script.turn_on service parameter

> Replaces `run_mode`.

Background/blocking behavior is now controlled by the `script.turn_on` service. The default (which also applies when calling a script "directly" -- `script.abc` -- or via the `script.toggle` service) is to be fully blocking. To run a script in the "background" a new `blocking` parameter to the `script.turn_on` service has been defined:

```yaml
- service: script.turn_on
  entity_id: script.abc
  data:
    blocking: false
    variables: ...
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

pnbruckner committed:
Add limit parameter to service call methods

pnbruckner committed:
Break out prep part of async_call_from_config for use elsewhere

pnbruckner committed:
Minor cleanup

pnbruckner committed:
Fix improper use of asyncio.wait

pnbruckner committed:
Fix state update

Call change listener immediately if its a callback

pnbruckner committed:
Fix exception handling and logging

pnbruckner committed:
Merge Script helper if_running/run_mode parameters into script_mode

- Remove background/blocking _ScriptRun subclasses which are no longer needed.

pnbruckner committed:
Add queued script mode

pnbruckner committed:
Disable timeout when making fully blocking script call

pnbruckner committed:
Don't call change listener when restarting script

This makes restart mode behavior consistent with parallel & queue modes.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -56,27 +69,23 @@
+    SCRIPT_MODE_QUEUE,
+    SCRIPT_MODE_RESTART,
 ]
+DEFAULT_SCRIPT_MODE = SCRIPT_MODE_LEGACY
```
We shouldn't specify a default. Let's make sure that we require the caller to specify a mode. That way we avoid surprises. It also allows us to print a warning ahead of changing the default from legacy -> blocking.

pnbruckner replied:
The `Script` helper class is used in a couple of other places as well (and who knows in how many custom integrations) and I wanted to make sure those continued to work as-is.

I'm handing the warning at the `script` and `automation` integration level. And rather than do it on a per script/automation basis I'm going to only issue one warning (per integration) that provides a list of the entities that haven't specified a mode. I thought one warning would be better than possibly many, many warnings (for people who have a lot of scripts and/or automations.)

Here's how it will look (from my script testing so far):
```
2020-03-03 13:26:42 WARNING (MainThread) [homeassistant.components.script] Script behavior has changed. To continue using previous behavior, which is now deprecated, add 'mode: legacy' to script(s): legacy_cancelable_not, legacy_cancelable_short, legacy_cancelable_long, new_run_new_no_wait_parallel_twice_direct, new_run_new_no_wait_parallel_twice_non_blocking_no_wait, new_run_new_no_wait_parallel_twice_non_blocking_wait, new_run_new_short_parallel_twice_direct, new_run_new_short_parallel_twice_non_blocking_no_wait, new_run_new_short_parallel_twice_non_blocking_wait, legacy_run_legacy_cancelable_not_twice_direct, legacy_run_legacy_cancelable_not_twice_turn_on, legacy_run_legacy_cancelable_short_twice_direct, legacy_run_legacy_cancelable_short_twice_turn_on, legacy_run_legacy_cancelable_long_twice_direct, legacy_run_legacy_cancelable_long_twice_turn_on, legacy_run_new_no_wait_parallel_twice_direct, legacy_run_new_no_wait_parallel_twice_turn_on, 1581351491735, 1581352147150, 1581355326399. See https://www.home-assistant.io/integrations/script/
```

balloob replied:
The only downside to the warning approach seems to be that we now force people to always specify a mode. We should make sure the frontend does this moving forward.

pnbruckner replied:
... or live with the warning (until it doesn't work anymore) which is what a lot of people seem to do. 😜

But seriously, we'll only force that until we remove legacy support. (Not sure what you're thinking along those lines.) After that I would expect we'd make the default something else reasonable, possibly depending on the integration. And then there will be no need for a warning because the default will be documented and nothing will be deprecated anymore.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
+    async def _async_call_service_step(self):
+        """Call the service specified in the action."""
+        domain, service, service_data = self._prep_call_service_step()
+        blocking = service_data.get(ATTR_BLOCKING, True)
```
This prevents us from passing `blocking` to the service that we're going to call. It also means that we can end up passing a variable to a service that is not valid for the service schema. We need to find a different way of passing `blocking` here. 

I also think that specifying `blocking` here is maybe out of scope of this PR? It seems like an unrelated change from specifying the run mode ?

pnbruckner replied:
Believe me, this is the area that caused me the most trouble and time. I tried at least three different approaches:

1. The original `run_mode` script helper parameter.
1. Modify the basic service call mechanism so that _any_ service could be called in the background.
1. Modify the script helper's service handling method to allow _any_ service to be called in the background.

All of these had practical issues (which are too difficult to explain in a short comment.)

I found this to be the most practical way to handle it. And it's in scope because this method has to use this information to determine what to set the call limit to. (The rest of the implementation is in the script integration, which I already have working.)

And this doesn't "pass" the parameter to any services. It only looks for it if it's there, and _then_ only uses it if the service is `script.turn_on`. (You'll notice it still sets `blocking=True` when it gets to the `hass.services.async_call` invocation.) I don't see how this can interfere with any other services. Again, it's not passing it to anything, it's only using it if it's there. (BTW, it's being added to the `script.turn_on` service's schema. Again, that's already in the script integration modifications I've already made and tested and will be in the next PR.)

balloob replied:
Ah I see it now. I didn't realize that it was only used inside the if-statement.

Why don't we just always call script domain without a limit ? If it's a non blocking run, it will just return very fast and we're ok?

pnbruckner replied:
Yes, that would work. I just assumed the limit would be useful, but I suppose that's probably really more for faulty integrations as opposed to core things. So, yeah, I can just disable the limit for all `script` services, which simplifies this a bit and decouples it from the `script.turn_on` service implemented in the `script` integration. Nice.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
+            )
+            for pending_task in pending:
+                if pending_task == task:
+                    # The scripts did not all finish. Stop them by turning them off.
```
We should be able to cancel our task and let asyncio do the work. We should not add custom logic because we can never support everything. Since scripts are shielded, they will not just stop half-way.

pnbruckner replied:
The intent is to stop scripts gracefully (i.e., between steps, or to stop the current step if it's a `delay`, `wait_template` or a call to another script.) And, yes, they are shielded, and if the task that calls `Script.asyn_run` is cancelled it will handle it and stop gracefully.

But this is not in the script or automation integration, so it doesn't have direct access to the corresponding entities. More to the point, it doesn't even necessarily have direct access to the tasks that called the `async_run` method of the potentially several `Script` objects. So instead it's using the service `script.turn_off` to tell those scripts / automation action sequences to stop. We are at the service call level here, so it's using another service call to effectively counter act those calls if the current script is stopped.

balloob replied:
I don't understand why we would need direct access to the corresponding entities? 

We're calling with blocking=True, so the cancellation should propagate all the way up (down?) to the script object that is being called.

pnbruckner replied:
The call to `hass.services.async_call` is indeed with `blocking=True` (and it needs to be that way for other reasons), but the `blocking` parameter of the `script.turn_on` service (which is different) will effectively make it non-blocking overall. So we'll come back from the call to `hass.services.async_call` possibly _before_ the run is done.

To put that another way, for the `script.turn_on` service, when the blocking call to `hass.services.async_call` returns depends on the _service's_ `blocking` _parameter_. If it's `True`, then the call will return when the entire script run has finished (or been stopped.) If it's `False`, then the call will return as soon as the script run _has started_ (and it's state has been updated to `'on'` to indicate it is running.)

Maybe I should change the name of the `script.turn_on` service's parameter to `background` and default it to `False`. Then again, based on the previous discussion, I'll be removing the need to look at the `script.turn_on` service's parameter here, so maybe the confusion will go away anyway.

balloob replied:
Well I don't think that we should try to stop scripts that we run in the background. If I am in a website and I open a link in another tab, closing the first tab doesn't close the second one. Same with starting threads in Python.

pnbruckner replied:
Not sure that's the same thing. The idea here is:
```yaml
script:
  a:
    sequence:
    - service: script.turn_on
      entity_id: script.b
      data:
        background: true
    - do_something_else_for_a_while
    - wait_template: "{{ is_state('script.b', 'off') }}"
    - blah_blah_blah
```
Now something starts `script.a`, but then decides that should be stopped so calls `script.turn_off` on it. `script.b` is just an implementation detail of `script.a`. If I want `script.a` to be stopped, then I want _everything it's doing_ to be stopped, including `script.b`.

Probably a better example is:
```yaml
script:
  a:
    mode: restart
    sequence:
    - ...
```
Or even:
```yaml
automation:
- ...
  mode: restart
  action:
  - ...
```

In these cases I want the script to restart for some reason or the automation is triggered again while the actions are still running. That should stop everything it's doing, including any scripts it called directly or indirectly, so it can all start over.

These are just a few simple examples, but it could get more involved with sub-scripts of sub-scripts. I shouldn't have to stop them all individually. That would become a maintenance nightmare.

balloob replied:
So first, this doesn't really belong in this PR. This seems like a new feature.

Second, since we support simultaneous executions of scripts, it might be that A stopping his own instance of script B by calling turn_off on script B, it is also stopping any other instance of script B that is running in parallel and could have been invoked by other parts of the system.

pnbruckner replied:
I think it's a natural result of having a restart feature, and I think it does need to be addressed as an inseparable piece of this overall feature set. If a script, or more likely an automation's action sequence, is restarted, it seems natural and expected that anything it does should get stopped so it can all start over again.

Yes, I agree that it is possible for a script to get used by more than one other script and/or automation, and therefore, yes, that would be a side effect -- i.e., one caller gets stopped and all invocations of the "used" script get stopped. I think that would apply equally well whether the script was called in a blocking mode or was started to run in the background. I do see that as a potential problem now that you point it out. But if it's a trade-off between the two behaviors  -- only stopping the outermost script, or stopping all called scripts, recursively, as well -- my opinion is the latter is what would be likely expected and needed. If someone wants a script to run completely independently, they can simply duplicate the script (possibly using YAML anchors.)

I suppose to do it in such a way there are no surprises, whenever one sequence (I mean a Script helper object) invokes another script (and here I mean a script integration entity), maybe it should keep a reference to the task that is executing it (whether in the "foreground" or "background") so it can be canceled (and like you suggested earlier, it can stop itself gracefully.) Maybe that's the best of both worlds.

balloob replied:
I think that we need a best of both world solutions. Having a script randomly stopped because another script got stopped is not easy to debug. It's a lot easier to debug if something continue to run. It's not possible to keep a reference to the task because we're just calling a service.  I suggest we keep this for a future PR and let's explore if we can use the context ID. But that _will_ get complicated and so we should not have it extend this PR.

pnbruckner replied:
You're completely right and I totally agree.

I'm not sure how I got on the subject of stopping a background script run. That's not what the code currently does or was intended to do. It is something that needs to be thought about and probably addressed, but that can be done later.

Also, you're right that the stopping of the service call to the script should be done via canceling the task. That should work and highlights the fact that I probably implemented the shielding at the wrong level.

I'll work on this and make another commit along these lines.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
-        except asyncio.CancelledError:
-            await self.async_stop()
-            raise
+            self._script._queue_lck.release()
```
Isn't this dangerous as you could end up releasing for another run?

pnbruckner replied:
The lock is shared by all runs of the _same script_. So if this run acquired the lock, it's safe (and the right thing to do) for it to release it, which will allow the next run to get it.

If, however, the script is stopped (i.e., the `self._stop.wait()` task finishes first), then _all_ runs of the script will also stop, too, so they they won't run anyway. I needed the try block to handle the case where the stop finished first.

So, bottom line, no, I believe this is correct.

pnbruckner replied:
Actually, now that I think about it some more, the stop flag is unique to the run, not the script. Although in most cases they will all get set because the script will stop all runs at the same time, I suppose it's possible for a run to get cancelled by itself, and in this case it will not have acquired the lock (if it's still in this piece of code.) So, yeah, you're probably right. I'll fix it. Thanks!

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
-    """Manage blocking Script sequence run."""
+    async def _async_run(self):
+        # pylint: disable=protected-access
+        _, pending = await asyncio.wait(
```
Please add a comment to explain what this logic does for future readers.

pnbruckner replied:
Will do.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -582,47 +684,60 @@ def __init__(
         self.name = name
         self._change_listener = change_listener
+        self._script_mode = script_mode
+        if logger:
```
All this logic, I feel like we should just make it required to pass in a logger. Let the instantiater take care of this. Our logs are not very helpful by making them relative to `helpers.script`.

pnbruckner replied:
I agree, but how do I do that without forcing all existing users to provide a logger (i.e., without adding it as a positional parameter)? I felt like that would have been a more disruptive change than just providing a reasonably default logger.

balloob replied:
There are only 2 users right now ?

pnbruckner replied:
17:
```
(venv) phil@GGHWFS1:~/repos/home-assistant (script-modes)$ grep -m 1 'Script(' homeassistant -r --include='*.py'
homeassistant/helpers/script.py:class _StopScript(Exception):
homeassistant/components/philips_js/media_player.py:    on_script = Script(hass, turn_on_action) if turn_on_action else None
homeassistant/components/lg_netcast/media_player.py:    on_action_script = Script(hass, on_action) if on_action else None
homeassistant/components/knx/__init__.py:        self.script = Script(hass, action, script_name)
homeassistant/components/kodi/media_player.py:            turn_on_action = script.Script(
homeassistant/components/wake_on_lan/switch.py:        self._off_script = Script(hass, off_action) if off_action else None
homeassistant/components/template/fan.py:        self._on_script = Script(hass, on_action)
homeassistant/components/template/lock.py:        self._command_lock = Script(hass, command_lock)
homeassistant/components/template/alarm_control_panel.py:            self._disarm_script = Script(hass, disarm_action)
homeassistant/components/template/light.py:        self._on_script = Script(hass, on_action)
homeassistant/components/template/cover.py:            self._open_script = Script(hass, open_action)
homeassistant/components/template/vacuum.py:        self._start_script = Script(hass, start_action)
homeassistant/components/template/switch.py:        self._on_script = Script(hass, on_action)
homeassistant/components/samsungtv/media_player.py:        on_script = Script(hass, turn_on_action)
homeassistant/components/automation/__init__.py:            action_script = script.Script(
homeassistant/components/webostv/media_player.py:    on_script = Script(hass, turn_on_action) if turn_on_action else None
homeassistant/components/intent_script/__init__.py:            conf[CONF_ACTION] = script.Script(
homeassistant/components/script/__init__.py:        self.script = Script(
```

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
 
-class _BackgroundScriptRun(_ScriptRun):
-    """Manage background Script sequence run."""
+        if limit is None:
```
Make this a guard clause. 

```python
if limit is not None:
    await coro
    return
```

pnbruckner replied:
Will do.

balloob commented on homeassistant/helpers/script.py:
```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
+        if domain == "script" and (
+            service == SERVICE_TURN_ON and blocking or service != SERVICE_TURN_OFF
+        ):
+            limit = None
```
This can be replaced by 

```suggestion
        if domain == "script" and service != SERVICE_TURN_OFF:
            limit = None
```

pnbruckner replied:
✔️

pnbruckner commented:
I've made the other changes already requested and tested them. I'll go ahead and make another commit so they can be looked over while we decide what to do about stopping/restarting invoked scripts.

pnbruckner committed:
Changes per review

- Call all script services (except script.turn_off) with no time limit.
- Fix handling of lock in _QueuedScriptRun and add comments to make it
  clearer how this code works.

pnbruckner committed:
Changes per review 2

- Move cancel shielding "up" from _ScriptRun.async_run to Script.async_run
  (and apply to new style scripts only.) This makes sure Script class also
  properly handles cancellation which it wasn't doing before.
- In _ScriptRun._async_call_service_step, instead of using script.turn_off
  service, just cancel service call and let it handle the cancellation
  accordingly.

balloob commented on homeassistant/core.py:
```diff
@@ -1230,7 +1219,7 @@ def call(
         try:
-            with timeout(SERVICE_CALL_LIMIT):
+            async with timeout(limit):
                 await asyncio.shield(self._execute_service(handler, service_call))
```
I wonder if this shield should be removed and instead the API endpoints that call this method should shield it. Otherwise we can't cancel a blocking script.

pnbruckner replied:
Hmm, I had forgotten this shield was here. It wasn't a problem before when I used the `script.turn_off` service to effectively cancel script runs. But I still agree that cancelling is the better way, and this shield _will_ prevent that from working. I guess the tests that have been written so far don't check this scenario, and I didn't re-run my ad hoc tests before committing this last set of changes. ☹️ I guess I need to beef up the tests (oh, and I still have to deal with the tests that wait on `asyncio.sleep`.)

What do you think the ramifications might be if we just remove this shield?

pnbruckner replied:
After thinking about this some more, and doing some experimenting, I realized that the shield was probably originally there to shield the coroutine from the timeout, but unfortunately for now it also shields cancellations from the caller. But that's actually very easily resolved. I replaced this:
```python
        try:
            async with timeout(limit):
                await asyncio.shield(self._execute_service(handler, service_call))
            return True
        except asyncio.TimeoutError:
            return False

```
with this:
```python
        task = self._hass.loop.create_task(self._execute_service(handler, service_call))
        try:
            async with timeout(limit):
                await asyncio.shield(task)
            return True
        except asyncio.TimeoutError:
            return False
        except asyncio.CancelledError:
            task.cancel()
            raise
```
That maintains the previous behavior (including the shield), but now allows cancellations from the caller to also cancel the service task. With this change all the tests passed as-is. (Basically it's just manually creating the task before calling `asyncio.shield` rather than letting `shield` create the task, so now it has a handle to the task it can use to cancel it.)

I also realized (via ad hoc testing) that the limit has more to do with _what is called_ than _who makes the call_. Specifically, no matter who makes the call, if a script is being called, the limit needs to be disabled. So I'm also moving the limit determination from the Script helper to this call by preceding the above with:
```python
        if domain == "script" and service != SERVICE_TURN_OFF:
            limit = None
        else:
            limit = SERVICE_CALL_LIMIT
```
and removing the `limit` parameter I had added to `call` and `async_call`.

The next commit will include these changes as well as the other fix I mentioned. I think they are necessary for this PR.

pnbruckner commented:
FYI, I did a rebase that I need to force push to resolve the conflicts from your PRs.

balloob commented:
This looks good ! We can merge this after conflicts resolved. Conflicts caused by me moving some constants to `config_validation.py`. 

Let's do the unshielding of the service call in core.py in a future PR.

pnbruckner commented:
I did some ad hoc testing and found a bug in the turn off sequence initiated by a cancellation. (Turning off was originally done via a service so I had overlooked this one.)

I want to commit that fix, then do a lot more testing (both with additional formal tests as well as ad hoc) before we merge this change. Give me at least a few more days.

Regarding removing the shield from `services.async_call`, I suppose it makes sense to do that separately since it has a potential for a large impact. I've been testing with that change and found at least a couple tests from other components that seem to be affected by the small timing change (i.e., probably one less await inside the call.)

pnbruckner committed:
Fix bugs

- Add missing call to change listener in Script.async_run
  in cancelled path.
- Cancel service task if ServiceRegistry.async_call cancelled.

pnbruckner commented:
Force push required due to rebase. See last commit for changes since last review.

pnbruckner commented:
I'd still like to do more testing, possibly adding more formal tests, before this PR is merged.

balloob commented on homeassistant/core.py:
```diff
@@ -1218,12 +1218,16 @@ def call(
         except asyncio.TimeoutError:
             return False
+        except asyncio.CancelledError:
+            task.cancel()
```
We can't do this. This is literally the only thing that shield does. This means you're removing shield here.

balloob replied:
FYI https://github.com/python/cpython/blob/89aa7f0ede1a11c020e83f24394593c577a61509/Lib/asyncio/tasks.py#L839

pnbruckner replied:
No, I don't think so.

If the `timeout` context manager does not expire before the task finishes the `async_call` returns `True`.

If, however, the `timeout` context manager does expire it does two things. First it tries to cancel the `await` _inside_ the `async with`. It does cancel the `await`, but the `shield` prevents it from canceling the task. It then raises a `TimeoutError` _outside_ the `async with` (not a `CancelledError`) which causes the `async_call` to return `False`. The task, however, continues to run, just like before. Nothing different yet.

Now, in the scenario where the task awaiting `async_call` is cancelled, that causes a `CancelledError` to be raised in `await asyncio.shield(task)`, and yes, that will not cancel the task because it is shielded, but then in the `except asyncio.CancelledError` clause the task is _directly_ cancelled (which we can do now because now we have a handle to it) before re-raising the `CancelledError`.

I've tested all four scenarios:

1. Script completes normally before the timeout.
1. Script completes normally but after the timeout.
1. Script is cancelled before the timeout.
1. Script is cancelled after the timeout.

All work as expected.

pnbruckner replied:
I guess in summary, yes, I'm effectively removing the shield, but only for when async_call is being cancelled, but I'm not for a timeout -- that stays the same as before.

balloob replied:
We shouldn't use a shield and then work around what does. It's confusing code and we can't have that in the core. 

Instead, can't we use a [`wait`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) with a timer and a task and pass `return_when=FIRST_COMPLETED`?

This is also a big change. We need to make sure that `websocket_api` and `api` integrations both call their services with a shield to avoid cancellation if the connection is dropped. We also probably need to revisit all places that process webhooks and call scripts to make sure it's shielded. Like, a lot more code needs to be cancellation aware that didn't had to be that before. I suggest we drop this change from this PR. It is going to balloon.

pnbruckner replied:
FWIW, it's not "working around" the shield. The shield is there for one reason, to shield the task from a cancel caused by the timeout. That is still needed. Rather it's adding handling for a cancellation that comes from outside the block of code rather than from inside it. Also I did at first implement it like you said but it created a much larger, and much less elegant, block of code, because wait is such a primitive there is a lot that needed to be duplicated that `timeout` and `shield` already did. In the end, IMHO, it was much more confusing. So I still disagree, but I guess I don't really care one way or the other, as long as it works.

But, more importantly, the more I work on this the more I'm discovering that this basic approach (to enhancing the script helper and its users) is running into, and breaking (or, at least, tripping over), a _lot_ of assumptions. Everywhere I turn I find something else that needs to be addressed, and unfortunately I'm still very new to this code base. And discussing all these issues is not really practical in this PR forum. I'm more than happy to continue working on this -- I'd really like to see it through -- but it's going to take _way_ more time than I had originally thought (even after assuming it wouldn't be easy.)

From here this can go three ways: we put this PR on hold and I take a side step to work on the basic changes to ServiceRegistry.async_call and all its ramifications (and I've already thought of more than we've discusses so far), ultimately in a separate PR, and then pick this one back up later, working both PRs much more slowly and methodically; I throw in the towel and you or someone else takes up the charge, using whatever you like from what I started; or we completely scrap this whole approach (reverting the previous PR) until you and/or someone else can figure out how to start it again a different way. Your choice.

balloob replied:
I think that you're on the right track here with scripts etc. I think that cancellation of a subscript is not a show-stopper issue. It's annoying that it will only cancel after the current script is done, but that's still an improvement over today.

The shield is not just there to cancel out the timeout. It's to keep it running, no matter if the `async_call` is canceled, either through a timeout or because the caller canceled. Because this code has been there since day 1, none of the service handling code deals with cancellations. 

I think that we should be able to merge this PR without removing the shield. The cancellation will just make the current script finish. We can improve on that later. I'm happy to help and think along etc. I think that it's a great improvement. I don't think that it's necessary to work in parallel or throw in the towel.

pnbruckner replied:
Ok, thanks, that's encouraging.

Yes, I see what you're saying about service handlers not handling cancellation. (I had to add that to the `script.turn_on` service handler.) I was only thinking from the other direction -- callers of `async_call`. Good point.

I'll continue slugging away. I'll probably go silent for a while as I try to work through the issues and come up with reasonable approaches to solving them. Besides the changes to `async_call`, there are still a few cleanup things to accomplish on this PR assuming we'll merge it without those changes. If I need help should I keep the discussion going here, or would you recommend another channel (discord, etc.)?

balloob replied:
This channel works, that way we can discuss it around parts of the code.

balloob replied:
btw I believe that small PRs keep things manageable. Even if we don't expose new functionality yet, we can already merge refactor PRs like this one, that shouldn't impact normal operational behavior.

pnbruckner replied:
Okey dokey. About to push two commits. The first backs out the last changes to homeassistant/core.py, and the second includes the minor fixes I mentioned. It also fairly overhauled tests/helpers/test_script.py, using Parametrize, using/adding common test functions, and I found a way to mock timeout so the tests don't have to wait for real time to elapse.

pnbruckner committed:
Revert last changes to ServiceRegistry.async_call

pnbruckner committed:
Minor Script helper fixes & test improvements

- Don't log asyncio.CancelledError exceptions.
- Make change_listener a public attribute.
- Test overhaul
  - Parametrize tests.
  - Use common test functions.
  - Mock timeout so tests don't need to wait for real time to elapse.
  - Add common function for waiting for script action step.

pnbruckner commented:
Assuming the tests pass (they did locally), and you don't find any other issues, then I think we can put this one to bed and I'll start working on the `ServiceRegistry.async_call` changes we discussed.

balloob commented:
Booooooooooom 🎉
